Docker.DotNet/DockerClient.cs

```diff
@@ -53,10 +56,75 @@ internal DockerClient(DockerClientConfiguration configuration, Version requested
             Miscellaneous = new MiscellaneousOperations(this);
             Networks = new NetworkOperations(this);
 
-            _client = new HttpClient(Configuration.Credentials.Handler, false);
+            ManagedHandler.StreamOpener opener;
+            var uri = Configuration.EndpointBaseUri;
+            switch (uri.Scheme.ToLowerInvariant())
+            {
+                case "npipe":
+                    if (Configuration.Credentials.IsTlsCredentials())
+                    {
+                        throw new Exception("TLS not supported over npipe");
+                    }
+
+                    var segments = uri.Segments;
+                    if (segments.Length != 3 || !segments[1].Equals("pipe/", StringComparison.InvariantCultureIgnoreCase))
+                    {
+                        throw new ArgumentException($"{Configuration.EndpointBaseUri} is not a valid npipe URI");
+                    }
+
+                    var serverName = uri.Host;
+                    var pipeName = uri.Segments[2];
+
+                    uri = new UriBuilder("http", pipeName).Uri;
+                    opener = async (string host, int port, CancellationToken cancellationToken) =>
+                    {
+                        // NamedPipeClientStream handles file not found by polling until the server arrives. Use a short
+                        // timeout so that the user doesn't get stuck waiting for a dockerd instance that is not running.
+                        int timeout = 100; // 100ms
+                        var stream = new System.IO.Pipes.NamedPipeClientStream(serverName, pipeName);
+#if NET45
+                        await Task.Run(() => stream.Connect(timeout), cancellationToken);
+#else
+                        await stream.ConnectAsync(timeout, cancellationToken);
+#endif
+                        return stream;
+                    };
+
+                    break;
+
+                case "tcp":
+                case "http":
+                    var builder = new UriBuilder(uri);
+                    builder.Scheme = configuration.Credentials.IsTlsCredentials() ? "https" : "http";
+                    uri = builder.Uri;
+                    opener = null;
+                    break;
+
+                case "https":
+                    opener = null;
+                    break;
+
+                //case "unix":
+                // TODO
+
+                default:
+                    throw new Exception($"Unknown URL scheme {configuration.EndpointBaseUri.Scheme}");
+            }
+
+            _endpointBaseUri = uri;
 
-            _defaultTimeout = _client.Timeout;
+            ManagedHandler handler;
+            if (opener != null)
+            {
+                handler = new ManagedHandler(opener);
```

Comment:
If you can create one with a default constructor and one with an opener why not just handle the null case in the constructor? That way you can have one statement here for callers.